doc/decisions/0_drafts/transformations.md

```diff
@@ -0,0 +1,152 @@
+
+For example, here is a configuration file with a hypothetical format:
+
+```
+/DISPLAY/BRIGHTNESS = 100
+```
+
+As can be seen, the keys are in UPPERCASE within the configuration file.
+In Elektra keys are case-sensitive.
+As operations on keysets such as `ksLookup` operate with _runtime_ data after the post-storage phase, `kdb get /DISPLAY/BRIGHTNESS` will fail.
+For Elektra, the key `/DISPLAY/BRIGHTNESS` does not exist, as the `rename` plugin transformed this into the lowercase `/display/brightness`.
+
+This leads to problems with the notification plugins.
+As notification plugins are executed after the post-storage phase of the `set` operation, they will receive a keyset with the already transformed keys.
+In this example, the notification plugins will receive all-UPPERCASE keys, and send out notifications with those all-UPPERCASE keys.
+An application listening to those notifications will not be able to query Elektra for those keys, as for Elektra those UPPERCASE keys do not exist.
+
+Apart from the problems with notifications, the way key name changing plugins work also breaks change tracking in plugins like `dbus`.
+This is because key names are read-only when they are contained in a `KeySet`.
+In order to change the name of a key, such a plugin has to create a new key with the changed name, and delete the key with the old name.
+The `dbus` plugin implements change tracking by checking the 'key needs sync' flag instead of comparing the values.
+As new keys by design have the 'key needs sync' flag set, the plugins that implement change tracking via the flag will always erroneously detect transformed keys as changed.
+
+### Observed problems with changing values (normalization)
+
+Suppose we have a plugin that changes the value of a key to the hard-coded value `1` during the storage phase of the `set` operation.
+If a plugin does change tracking, this will lead to false positives.
+
+```
+user:/limits/openfiles = 1
+```
+
+If the user changes the value, e.g. using `kdb set user:/limits/openfiles 23`, plugins will observe the new value `23`.
+However, the value-changing plugin will reset that value back to `1`.
+So in practice, the configuration has not been changed.
+Plugins relying on change tracking plugins (e.g. notification plugins) will however think that it has.
+
+### Observed problems with adding and removing metadata
+
+This is a bit of a mixture between changing key names and changing values.
+
+## Constraints
+
+1.
+2.
+3.
+
+## Assumptions
+
+1. We want to enable some kind of transformations
+2. False positives for change tracking algorithms are only a minor problem.
+3. There is no reason to modify or delete existing `meta:/` keys.
+4. Newly generated `meta:/...` keys can
+   - either stay and get permanently stored during `kdbSet`
+   - or be written as `meta:/generated/...`.
```

Comment:
Isn't this the internal we already have?

Reply:
Similar semantics, but a separate prefix of generated data makes more sense IMO. Also `meta:/internal` is split by plugin and each `meta:/internal/<plugin>/foo` should only be used by `<plugin>` and ignored by everything else (at least that's how I understand it).